### PR TITLE
Add retain functions to Registry

### PR DESCRIPTION
**Pull request by kathoum**
This adds public functions retain_counters, retain_gauges, retain_histograms, that work like HashMap::retain

### CODE REVIEW

**Comment by kathoum** [comment]
Design-wise, I'm not sure about the choice on the function signature:
`FnMut(&K, &mut S::Counter) -> bool` or `FnMut(&K, &S::Counter) -> bool`

I went with `&mut` because there's already a write lock, so the code has exclusive access to the underlying `S::Counter`, and it shouldn't make much difference in the common case that `Counter = Arc<impl CounterFn>`.

But I understand if we prefer not allowing `mut` access to objects inside the storage.

**Comment by kathoum** [comment]
PR updated, changed the visitors to accept immutable references

File: metrics-util/src/registry/mod.rs [hunk 1]
```diff
@@ -456,6 +498,15 @@ mod tests {
...
 
+        registry.get_or_create_counter(&Key::from_name("baz"), |_| ());
+
+        let mut n = 0;
+        registry.retain_counters(|k, _| {
+            n += 1;
+            k.name().starts_with("foo")
+        });
+        assert_eq!(n, 2);
+
```
Comment (tobz) [inline metrics-util/src/registry/mod.rs:509]: I think you'd want to copy the assert from lines 512-513 and stick them after adding the `baz` counter, to show that we now have two counters in the registry. After that, do the same thing after the `retain_counters` call to show it goes back down to one counter after it deletes `baz`.

**Review by tobz — CHANGES_REQUESTED** [review]

**Review by tobz — APPROVED** [review]
Looks good to me. 👍🏻

**Comment by tobz** [comment]
Released in `metrics-util@v0.16.3`.

Thanks again for your contribution!

**Pull request merged**